### PR TITLE
Vendor  service converter liquibase

### PR DESCRIPTION
**Pull request by OlegKandalov**

### CODE REVIEW

File: src/main/java/com/exadel/discount/platform/model/dto/VendorDto.java [hunk 1]
```diff
@@ -16,7 +16,12 @@
     private String name;
     @Size(min = 50, max = 200)
     private String description;
+    private String email;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/model/dto/VendorDto.java:19]: use `@Email` to validate email addresses

We probably should also add `contactPhone`. @nkharitoniuk what do you think?

File: src/main/java/com/exadel/discount/platform/model/Vendor.java [hunk 2]
```diff
@@ -25,11 +27,17 @@
     @Size(min = 50, max = 200)
     @Column(name = "v_description")
     private String description;
+    @Column(name = "v_email")
+    private String email;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/model/Vendor.java:31]: use `@Email` to validate email addresses

File: src/main/resources/liquibase/21.06.21_oleg_create_vendor.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+            <column defaultValueComputed="uuid_generate_v4()" name="v_id" type="UUID">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="pk_vendor"/>
+            </column>
+            <column name="v_name" type="varchar(50)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="v_description" type="varchar(200)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="v_email" type="varchar(100)"/>
```
Comment (ohalushchak-exadel) [inline src/main/resources/liquibase/21.06.21_oleg_create_vendor.xml:17]: 256 (RFC 2821)

File: src/main/resources/liquibase/21.06.21_oleg_create_vendor.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+            <column name="v_image" type="varchar(100)"/>
```
Comment (ohalushchak-exadel) [inline src/main/resources/liquibase/21.06.21_oleg_create_vendor.xml:18]: Image url can be easily longer than that.  Make it 256

File: src/main/resources/liquibase/21.06.21_oleg_create_vendor_location.xml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                <constraints nullable="false"/>
+            </column>
+            <column name="vl_address_line" type="varchar(50)"/>
+            <column name="vendor_id" type="uuid">
+                <constraints nullable="false"/>
+            </column>
+            <column name="vl_deleted" type="boolean"/>
+        </createTable>
+
+        <addForeignKeyConstraint baseTableName="vendor_location"
```
Comment (ohalushchak-exadel) [inline src/main/resources/liquibase/21.06.21_oleg_create_vendor_location.xml:24]: Move this to a separate changeset withing the file. One change - one changeset

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    public void update(UUID id, VendorDto vendorDto) {
+        Vendor vendor = vendorRepository.findById(id).orElseThrow();
+        vendor = converter.dtoToEntity(vendorDto);
+        vendorRepository.save(vendor);
+    }
+
+    @Override
+    public void toArchive(UUID id) {
+        Vendor vendor = vendorRepository.findById(id).orElseThrow();
+        vendorRepository.deleteById(id);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:48]: What happens to the vendor's locations here? Cascade deletes in the Vendor entity or delete locations by vendor Id. Same will apply to the discounts later on.

File: src/main/java/com/exadel/discount/platform/converter/VendorConverter.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import com.exadel.discount.platform.model.Vendor;
+import com.exadel.discount.platform.model.dto.VendorDto;
+import org.modelmapper.ModelMapper;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Component
+public class VendorConverter {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/converter/VendorConverter.java:12]: VendorMapper maybe?

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public List<VendorDto> getAll() {
+        return converter.mapList(vendorRepository.findAll(), VendorDto.class);
+    }
+
+    @Override
+    public VendorDto save(VendorDto vendorDto) {
+        vendorRepository.save(converter.dtoToEntity(vendorDto));
+        return vendorDto;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:29]: return what you have saved, not what you have received

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public VendorDto getById(UUID id) {
+        Vendor vendor = vendorRepository.findById(id).orElseThrow();
+        return converter.entityToDto(vendor);
+    }
+
+    @Override
+    public void update(UUID id, VendorDto vendorDto) {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:39]: return saved/updated Vendor

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        Vendor vendor = vendorRepository.findById(id).orElseThrow();
+        vendor = converter.dtoToEntity(vendorDto);
+        vendorRepository.save(vendor);
+    }
+
+    @Override
+    public void toArchive(UUID id) {
+        Vendor vendor = vendorRepository.findById(id).orElseThrow();
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:47]: use vendorRepository.existsById() instead

File: src/main/java/com/exadel/discount/platform/model/VendorLocation.java [hunk 11]
```diff
@@ -30,7 +30,7 @@
     @Column(name = "vl_address_line")
     private String addressLine;
 
-    @ManyToOne(fetch = FetchType.LAZY)
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.REMOVE)
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/model/VendorLocation.java:33]: Vendor will be deleted if one of his locations is deleted. 

Cascade should be specified in the owning entity - Vendor, not here.

File: src/main/java/com/exadel/discount/platform/exception/BadIdException.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package com.exadel.discount.platform.exception;
+
+public class BadIdException extends Exception {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/exception/BadIdException.java:3]: This should be a RuntimeException. And since you're using it as an indicator that there's no such element in the database, name in 'NotFoundException' or something

File: src/main/java/com/exadel/discount/platform/service/interfaces/CrudService.java [hunk 13]
```diff
@@ -8,6 +10,6 @@
     List<T> getAll();
     T save(T t);
     T getById(UUID id);
-    void update(UUID id, T t);
-    void toArchive(UUID id);
+    T update(UUID id, T t);
+    void toArchive(UUID id) throws BadIdException;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/interfaces/CrudService.java:14]: remove 'throws' after you make it RuntimeException

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 14]
```diff
@@ -15,36 +16,41 @@
...
     public VendorDto save(VendorDto vendorDto) {
-        vendorRepository.save(converter.dtoToEntity(vendorDto));
-        return vendorDto;
+        Vendor vendor = vendorRepository.save(mapper.dtoToEntity(vendorDto));
+        return mapper.entityToDto(vendor);
     }
 
     @Override
     public VendorDto getById(UUID id) {
         Vendor vendor = vendorRepository.findById(id).orElseThrow();
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:35]: throw your new BadIdException (NotFoundException) here

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 15]
```diff
@@ -15,36 +16,41 @@
...
-        return converter.entityToDto(vendor);
+        return mapper.entityToDto(vendor);
     }
 
     @Override
-    public void update(UUID id, VendorDto vendorDto) {
+    public VendorDto update(UUID id, VendorDto vendorDto) {
         Vendor vendor = vendorRepository.findById(id).orElseThrow();
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:41]: throw your new BadIdException (NotFoundException) here

File: src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java [hunk 16]
```diff
@@ -15,36 +16,41 @@
...
     @Override
-    public void toArchive(UUID id) {
-        Vendor vendor = vendorRepository.findById(id).orElseThrow();
-        vendorRepository.deleteById(id);
+    public void toArchive(UUID id) throws BadIdException {
+        boolean isExists = vendorRepository.existsById(id);
+        if (isExists) {
+            vendorRepository.deleteById(id);
+        } else {
+            throw new BadIdException("object with this id was not found");
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorServiceImpl.java:53]: be more specific about what went wrong with what data and why. Which object? Which id?

File: src/main/java/com/exadel/discount/platform/model/VendorLocation.java [hunk 17]
```diff
@@ -17,17 +17,17 @@
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/model/VendorLocation.java:17]: Should it also have `@SQLDelete`?

**Review by ohalushchak-exadel — APPROVED** [review]

**Review by nkharitoniuk — APPROVED** [review]
